_posts/2019-06-25-building-a-mono-repo.markdown

```diff
@@ -0,0 +1,66 @@
+* Any of the above, but in repsonse to a dependent project being "built" (for that project's definition of build)?
+* Or, God forbid, all of the above?
+
+I pinged Dmitriy on Twitter after his initial post and he confirmed that he needed all of the above. You have my condolences, Dmitriy.
+
+
+# Ideas Informed by Experience
+
+I won't endorse any particular approach because:
+1. The "right" approach is more likely to be dictated by the nature of the software you're building than anything else, and
+1. You didn't think I was going to make this easy on you, did you?
+
+I will however run through three approaches I've used in the past and explain the pros and cons of each.
+
+#### 1. Bash Your Way Through It
+
+Start with a simple bash script that understands project dependencies, looks at the diff of the change set in question, and infers what needs to be built in what order. It grows and grows, and six months later you realize you've written your own build system. Weeping ensues.
+
+**_Pros_**: Your build is defined in your repo and so in theory should be owned by the whole team. Anyone can add a new project and configure its dependencies (hopefully).
+
+**_Cons_**: Loss of build transparency. Questions like "*how often does project Y get built*" or "*how long does project X take to build on average*" are difficult to answer. They likely require parsing build logs and trying to regex out the time to build project Y. Assuming you thought to print that info to begin with.
+
+You are frequently having to remind that one person who answers these questions with data from your Jenkins monojob that "*we build something different on each run*" so "*those numbers are basically meaningless*", and "Seriously Garry, we just talked about this last week*".
+
+#### 2. Directed A-Sickening Graph
+Give each of your projects their own Jekins job. Use the path regex option to restrict builds from triggering unless said project has been modified. Create dependencies between each of these jobs that models the Directed Acyclic Graph of your project dependencies. Setup the [dependency graph plugin](https://wiki.jenkins.io/display/JENKINS/Dependency+Graph+View+Plugin) after a small change to a `README` file unexpectedly causes every project to be rebuilt and deployed. Then react in horror as it shows you the monsterous dependnecy graph you're created. At least with a shell script no can actually see the mess.
+
+
+**_Pro_**: Transparency! For better or worse you'll always know when, where, and why each build has been triggered.
+
+**_Cons_**: You've taken complexity that will exist no matter what and moved it from your repo into Jenkins configuration that no one really understands... except maybe you? Adding a new project results in a slew of Jiras for Jenkins setup that are... wait, why are these assigned me? I thought we agree'd people would self service? Wait, I'm the only one who knows how to set these up? Ug, fine.
```

Comment:
"a Jenkins config" or "Jenkins config" not sure which is idiomatic here.

Reply:
Jenkins configuration. This the abstract concept of it living in configuration rather than code. Going the other way around would be like saying "moving it into a code".

I think the article `a` would have to mean that "jenkins configuration" is a specific unit. In Jenkins parlance you'd say "a Jenkins project" for a specific configured build.